cpp/include/kvikio/file_handle.hpp

```diff
@@ -497,28 +497,36 @@ class FileHandle {
   }
 
   /**
-   * @brief Reads specified bytes from the file into the device memory.
+   * @brief Reads specified bytes from the file into the device memory asynchronously.
    *
-   * This API reads size bytes asynchronously from the file into device memory writing
-   * to a specified offset using GDS functionality. The API works correctly for unaligned
-   * offset and data sizes, although the performance is not on-par with aligned read.
-   * This is an asynchronous call and will be executed in sequence for the specified stream.
+   * This is an asynchronous version of `.read()`, which will be executed in sequence
+   * for the specified stream.
    *
-   * @note For the `devPtr_offset`, if data will be read starting exactly from the
-   * `devPtr_base` that is registered with `buffer_register`, `devPtr_offset` should
-   * be set to 0. To read starting from an offset in the registered buffer range,
-   * the relative offset should be specified in the `devPtr_offset`, and the
-   * `devPtr_base` must remain set to the base address that was used in the
-   * `buffer_register` call.
+   * The arguments have the same meaning as in `.read()` but some of them are deferred. That is,
+   * the values of `size`, `file_offset` and `devPtr_offset` will not be evaluated until execution
+   * time. Notice, this behavior can be changed using cuFile's cuFileStreamRegister API.
    *
    * @param devPtr_base Base address of buffer in device memory. For registered buffers,
    * `devPtr_base` must remain set to the base address used in the `buffer_register` call.
-   * @param size Size in bytes to read.
-   * @param file_offset Offset in the file to read from.
-   * @param devPtr_offset Offset relative to the `devPtr_base` pointer to read into.
-   * This parameter should be used only with registered buffers.
-   * @param bytes_read number of bytes that were successfully read.
-   * @param stream associated stream for this I/O.
+   * @param size Pointer to size in bytes to read. If the exact size is not known at the time of I/O
+   * submission, then you must set it to the maximum possible I/O size for that stream I/O. Later
+   the
+   * actual size can be set prior to the stream I/O execution.
+   * @param file_offset Pointer to offset in the file from which to read. Unless otherwise set using
+   cuFileStreamRegister API, this value will not be evaluated until execution time.
+   * @param devPtr_offset Pointer to the offset relative to the bufPtr_base pointer from which to
+   * write. Unless otherwise set using cuFileStreamRegister API, this value will not be evaluated
+   * until execution time.
+   * @param bytes_read Pointer to the bytes read from file. This pointer should be a non-NULL value
+   * and *bytes_read set to 0. The bytes_read memory should be allocated with cuMemHostAlloc/malloc/
+   * mmap or registered with cuMemHostRegister.
```

Comment:
I know you have just copied the cufile docs here, but this sentence makes no sense to me. AIUI, just plain `malloc` and `mmap` are completely unknown to the driver. So does this mean use `cuMemHostAlloc` or `cuMemAlloc` or `cuMemMap`? Or am I fully even more confused than I thought?

Reply:
I guess it means that it has to be allocated on the heap (not stack)?

Reply:
That's correct.